### DISCUSSION TITLE
List of pnpm packages for programmatic usage

### DISCUSSION THREAD

**Discussion by jcao02** [pnpm/pnpm#7699] [General]
Hi,

In my company we are using `pnpm` as the default package manager in across 45+ web applications. The tools works like a charm and the speed is really impressive (kudos!!)

We are leveraging the workspace feature to host our web applications inside monorepositories and we have enabled smart CI checks using the `--filter` flag of the CLI. However, we are trying to go further with this and leverage the tool for many other use cases. Some examples:
1. Outdated dependencies analysis: check the amount of outdated dependencies, their diff with the current version, their penetration in the codebase, etc
2. Ownership analysis: check CODEOWNERS vs packages in workspace

Most of the use-cases are automated inside Github workflows and we are currently creating a child process to use the CLI, then parse the output and bring it back to NodeJS runtime for manipulation -- there's a performance overhead involved and sometimes the CLI behaviour is not intuitive in NodeJS (e.g. `pnpm outdated` returns 1 status code if there are outdated dependencies, but in NodeJS I'd expect a function `getOutdatedDependencies` to just return a list without failing).

Is there a list of official pnpm packages that mirror/extend the CLI commands? We are using some packages already like `@pnpm/find-workspace-dir` and `@pnpm/workspace.find-packages`, but we tried to use `@pnpm/outdated` and it looked more like a private package because in order to use it, we'd need a lot knowledge of the codebase implementation details.

If there is no list of packages, is there any plan to create one? I think this would scale the adoption even further because consumers will be able to use general purpose languages (i.e. TypeScript/JavaScript) to implement useful tools for the community.

Thanks and keep the good work!

**Comment by zkochan** [comment]
There's no such list. The issue with this is that it would increase maintenance burden to properly make the packages "public". They are public but with no guarantee.

I would suggest to use the CLI directly, if possible. Some commands support the `--json` flag which makes the output easier to parse.

I use many packages from pnpm in the bit repository but I know the packages, so I don't need help from maintainers or documentation.

**Reply by jcao02** [reply]
I see. Thanks for the reply. I'll go ahead with the CLI then. Just for visibility, the `--json` flag for outdated dependencies is not working as expected https://github.com/pnpm/pnpm/issues/7693

**Comment by hilja** [comment]
Have you looked into https://github.com/raineorshine/npm-check-updates? It's a neat program for dealing with outdated packages. It has a ton of options, I haven’t explored it very deeply/ I just use the interactive mode to update deps in a monorepo:

```json
{
  "scripts": {
    "check": "pnpm dlx npm-check-updates -i --format group -p pnpm --cache --workspaces --root"
  }
}
```

**Reply by jcao02** [reply]
I haven't! Thanks for the visibility -- I'll take a look.
